File: src/main/scala/com/kapilkoju/autopos/party/api/CustomerApi.scala
package com.kapilkoju.autopos.party.api

import java.util
import javax.validation.Valid

import com.codahale.metrics.annotation.Timed
import com.kapilkoju.autopos.party.api.CustomerApi.baseUrl
import com.kapilkoju.autopos.party.domain.Customer
import com.kapilkoju.autopos.party.service.{CustomerRepo, CustomerService}
import com.kapilkoju.autopos.web.rest.util.{HeaderUtil, PaginationUtil}
import org.springframework.data.domain.Pageable
import org.springframework.http.HttpStatus.{NOT_FOUND, OK}
import org.springframework.http.{HttpStatus, ResponseEntity}
import org.springframework.util.Assert
import org.springframework.web.bind.annotation._

@RestController
@RequestMapping(Array(baseUrl))
class CustomerApi(repo: CustomerRepo, customerService: CustomerService) {

  final val entityName = "customer"

  @GetMapping(value = Array("/{id}"))
  @Timed
  def get(@PathVariable("id") id: Long): ResponseEntity[_ <: Customer] = {
    customerService.getCustomer(id) match {
      case Some(customer) => new ResponseEntity(customer, OK)
      case None => new ResponseEntity(NOT_FOUND)
    }
  }

  @GetMapping
  @Timed
  def getAll(pageable: Pageable): ResponseEntity[util.List[Customer]] = {
    val page = repo.findAll(pageable)
    val headers = PaginationUtil.generatePaginationHttpHeaders(page, baseUrl)
    new ResponseEntity(page.getContent, headers, HttpStatus.OK)
  }

  @PostMapping
  @Timed
  def save(@RequestBody @Valid customer: Customer): ResponseEntity[Customer] = {

    if (customer.getId != null) {
      val failureHeaders = HeaderUtil.createFailureAlert(entityName,
        "idexists", s"A new $entityName cannot already have an ID")
      return new ResponseEntity[Customer](failureHeaders, HttpStatus.BAD_REQUEST)
    }

    val savedEntity = repo.save(customer)
    ResponseEntity
      .ok
      .headers(HeaderUtil.createEntityCreationAlert(entityName, savedEntity.getId.toString))
      .body(savedEntity)
  }


  @PutMapping(value = Array("/{id}"))
  @Timed
  def update(@PathVariable("id") id: Long, @RequestBody @Valid customer: Customer): ResponseEntity[Customer] = {

    Assert.isTrue(customer.getId == id)

    val updatedEntity = repo.save(customer)
    ResponseEntity
      .ok
      .headers(HeaderUtil.createEntityUpdateAlert(entityName, customer.getId.toString))
      .body(updatedEntity)
  }

  @PutMapping
  @Timed
  def updateEntity(@RequestBody @Valid customer: Customer): ResponseEntity[Customer] = {
    Assert.isTrue(customer.getId != null)
    update(customer.getId, customer)
  }


  @DeleteMapping(value = Array("/{id}"))
  @Timed
  def delete(@PathVariable("id") id: Long): ResponseEntity[Void] = {
    repo.delete(id)
    ResponseEntity
      .ok
      .headers(HeaderUtil.createEntityDeletionAlert(entityName, id.toString))
      .build()
  }
}

object CustomerApi {
  final val baseUrl = "/api/customers"
}

File: kotlang/src/main/kotlin/com/kapilkoju/autopos/party/api/CustomerApi.kt
package com.kapilkoju.autopos.party.api

import com.codahale.metrics.annotation.Timed
import com.kapilkoju.autopos.party.domain.Customer
import com.kapilkoju.autopos.party.service.CustomerRepo
import com.kapilkoju.autopos.party.service.CustomerService
import com.kapilkoju.autopos.web.rest.util.HeaderUtil
import com.kapilkoju.autopos.web.rest.util.PaginationUtil
import org.springframework.data.domain.Pageable
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.util.Assert
import org.springframework.web.bind.annotation.*
import javax.validation.Valid

@RestController
@RequestMapping(CustomerApi.baseUrl)
class CustomerApi(private val repo: CustomerRepo, private val customerService: CustomerService) {


    @GetMapping("{id}")
    @Timed
    fun get(@PathVariable("id") id: Long): ResponseEntity<Customer> {
        val customer = customerService.getCustomer(id)
        return if (customer != null) {
            ResponseEntity(customer, HttpStatus.OK)
        } else {
            ResponseEntity(HttpStatus.NOT_FOUND)
        }
    }

    @GetMapping
    @Timed
    fun getAll(pageable: Pageable): ResponseEntity<List<Customer>> {
        val page = repo.findAll(pageable)
        val headers = PaginationUtil.generatePaginationHttpHeaders(page, baseUrl)
        return ResponseEntity(page.content, headers, HttpStatus.OK)
    }

    @PostMapping
    @Timed
    fun save(@RequestBody @Valid customer: Customer): ResponseEntity<Customer> {

        if (customer.getId() != null) {
            val failureHeaders = HeaderUtil.createFailureAlert(entityName,
                    "idexists", "A new $entityName cannot already have an ID")
            return ResponseEntity(failureHeaders, HttpStatus.BAD_REQUEST)
        }

        val savedEntity = repo.save(customer)
        return ResponseEntity
                .ok()
                .headers(HeaderUtil.createEntityCreationAlert(entityName, savedEntity.getId().toString()))
                .body(savedEntity)
    }


    @PutMapping("/{id}")
    @Timed
    fun update(@PathVariable("id") id: Long, @RequestBody @Valid customer: Customer): ResponseEntity<Customer> {

        Assert.isTrue(customer.getId() == id, "id of customer to be updated should not be null")

        val updatedEntity = repo.save(customer)
        return ResponseEntity
                .ok()
                .headers(HeaderUtil.createEntityUpdateAlert(entityName, customer.getId().toString()))
                .body(updatedEntity)
    }

    @PutMapping
    @Timed
    fun updateEntity(@RequestBody @Valid customer: Customer): ResponseEntity<Customer> {
        Assert.isTrue(customer.getId() != null, "id of customer to be updated should not be null")
        return update(customer.getId()!!, customer)
    }


    @DeleteMapping("/{id}")
    @Timed
    fun delete(@PathVariable("id") id: Long): ResponseEntity<Void> {
        repo.delete(id)
        return ResponseEntity
                .ok()
                .headers(HeaderUtil.createEntityDeletionAlert(entityName, id.toString()))
                .build()
    }

    companion object CustomerApi {
        const val baseUrl = "/api/customers"
        const val entityName = "customer"
    }
}

